enhancements/machine-api/aws-imds-v2-support.md

```diff
@@ -0,0 +1,260 @@
+At this point, IMDSv2 enablement might be done manually via AWS console, but providing such settings on Machine API is a quite loud request from our end-users.
+
+### Goals
+
+- provide mechanisms to enable/enforce IMDSv2 for newly created machines via Machine API
+- provide mechanisms to enable/enforce IMDSv2 for master machines during installation procedure for newly created clusters
+
+### Non-Goals
+
+- perform automated migration between using IMDSv1 and IMDSv2 for existing infrastructure
+
+## Proposal
+
+ - Add new options into install-config for Othe CP installation program which will be passed as metadata service settings during virtual machines creation
+ - Extend [AWSMachineProviderConfig](https://github.com/openshift/api/blob/master/machine/v1beta1/types_awsprovider.go#L12) by adding metadata service parameters
+ - Pass metadata service parameters to AWS during new machines creation
+
+### User Stories
+
+- As an OpenShift cluster administrator, I would like to enforce IMDSv2 usage during an OCP cluster deployment procedure.
+- As an OpenShift cluster administrator, I would like to enforce IMDSv2 usage for newly created machines in my cluster via MachineSets configuration.
+- As an OpenShift cluster administrator, I would like to change IMDS interaction settings for newly created machines in my cluster.
+
+### API Extensions
+
+#### AWSMachineProviderConfig changes
+
+MachineAPI design mostly replicates parameters which [AWS API](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_InstanceMetadataOptionsRequest.html) consumes.
+
+```go
+type AWSMachineProviderConfig struct {
+    // Existing fields will not be modified
+    ...
+	MetadataServiceOptions *MetadataServiceOptions `json:"metadataServiceOptions,omitempty"`
+}
+```
+
+```go
+
+type InstanceMetadataHttpTokens string
+
+const (
+    // InstanceMetadataHttpTokensRequired enforces sending of a signed token header with any instance metadata retrieval requests
+    InstanceMetadataHttpTokensRequired = "required"
+    // InstanceMetadataHttpTokensOptional allows IMDSv1 usage along with IMDSv2
+    InstanceMetadataHttpTokensOptional = "optional"
+)
+
+// MetadataServiceOptions defines the options available to a user when configuring
+// Instance Metadata Service (IMDS) Options.
+type MetadataServiceOptions struct {
+    // HttpTokens is the state of token usage for instance metadata requests from this host.
+    // Effectively enforces IMDSv2 if set to required.
+    // Default: optional
+    // +kubebuilder:validation:Enum=required;optional
+    // +optional
+    HttpTokens InstanceMetadataHttpTokens `json:"httpTokens,omitempty"`
+    // HttpPutResponseHopLimit is the desired HTTP PUT response hop limit for instance metadata requests.
+    // The larger the number, the further instance metadata requests can travel.
```

Comment:
What does this mean in real terms for an end user

Reply:
Number of proxies/network hops between host and metadata service. TTL in other words.

Literally 

> HttpPutResponseHopLimit is the desired HTTP PUT response hop limit for instance metadata requests.

Reply:
Why would our end user want to configure this? What effect does it have to them? Do we need to expose this option?

Reply:
I think we need to explain this further because I'm still not really understanding it, is this a security feature? Why would I want to turn this on? Why would I want a smaller number vs a larger number? What happens if this value isn't provided?

Reply:
Expanded.